pysd/py_backend/output.py

```diff
@@ -0,0 +1,519 @@
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+
+        Returns
+        -------
+        None
+
+        """
+        self.ds = nc.Dataset(self.out_file, "w")
+
+        # defining global attributes
+        self.ds.description = "Results for simulation run on" \
+            f"{t.ctime(t.time())} using PySD version {__version__}"
+        self.ds.model_file = model.py_model_file
+        self.ds.timestep = f"{model.components.time_step()}"
+        self.ds.initial_time = f"{model.components.initial_time()}"
+        self.ds.final_time = f"{model.components.final_time()}"
+
+        # creating variables for all model dimensions
+        for dim_name, coords in model.subscripts.items():
+            coords = np.array(coords)
+            # create dimension
+            self.ds.createDimension(dim_name, len(coords))
+            # length of the longest string in the coords
+            max_str_len = len(max(coords, key=len))
+
+            # create variable
+            # TODO: check if the type could be defined otherwise)
+            var = self.ds.createVariable(
+                dim_name, f"S{max_str_len}", (dim_name,))
+            # assigning values to variable
+            var[:] = coords
+
+        # creating the time dimension as unlimited
+        self.ds.createDimension("time", None)
+
+        # creating variables in capture_elements
+        self.__create_ds_vars(model, capture_elements)
+
+    def update(self, model, capture_elements):
+        """
+        Writes values of variables in capture_elements in the netCDF4 Dataset.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+
+        Returns
+        -------
+        None
+        """
+        for key in capture_elements:
+
+            comp = getattr(model.components, key)
+            comp_vals = comp()
+
+            if "time" in self.ds[key].dimensions:
+                if isinstance(comp_vals, xr.DataArray):
+                    self.ds[key][self.step, :] = comp_vals.values
+                elif isinstance(comp_vals, np.ndarray):
+                    self.ds[key][self.step, :] = comp_vals
+                else:
+                    self.ds[key][self.step] = comp_vals
+            else:
+                try:  # this issue can arise with external objects
+                    if isinstance(comp_vals, xr.DataArray):
+                        self.ds[key][:] = comp_vals.values
+                    elif isinstance(comp_vals, np.ndarray):
+                        if comp_vals.size == 1:
+                            self.ds[key][:] = comp_vals
+                        else:
+                            self.ds[key][:] = comp_vals
+                    else:
+                        self.ds[key][:] = comp_vals
```

Comment:
I also don't get which object can have time as dimension or when the try/except arises

Reply:
> I also don't get which object can have time as dimension or when the try/except arises

The `time` dimension is used when the variable has cache step, and at each iteration, the index of time increases by 1.

I had to include the try/ except for an external object that was partially defined (only some coordinates, but not all), in the WILIAM model:

`
WEIGHT_OF_4W_VEHICLES_IN_DEMAND_OF_THE_SECTOR[Transport_Equipment_M]=
	GET_DIRECT_CONSTANTS('model_parameters/economy/economy.xlsx', 'World', 'weitgh_of_4w_in_demand_of_Transport_Equip'\
		) ~~|`

`
WEIGHT_OF_4W_VEHICLES_IN_DEMAND_OF_THE_SECTOR[Sale_Maintenance_and_Repair_of_Motor_Vehicles_anda_Motorcycles_Retail_Sale_of_fuel_M\
		]=
	GET_DIRECT_CONSTANTS('model_parameters/economy/economy.xlsx', 'World', 'weitgh_of_4w_in_demand_of_Sale_Maintenan'\
		)
	~	DMNL
	~	Weight of 4w vehicles over total sector demand
	|
`

Note that  `Transport_Equipment_M` and `Sale_Maintenance_and_Repair_of_Motor_Vehicles_anda_Motorcycles_Retail_Sale_of_fuel_M` are coordinates of the INDUSTRY_SECTORS_I subscript, which has many other coordinates.

When this happens, the shape of the DataArray does not coincide with the model.component subscripts. I suspect this should be prevented before reaching this point.

Reply:
But, as I understand, then the second block should only be run when the model is finished, as is it for the constant variables.

About the error with the subscripts, it may be necessary to modify a little bit the builder, but I think it is important to try giving always the same dimensions with nan values in the other subscripts.  I could take care of that.